File: 2023-Xtreme-Xmas-Code-Maiden-Voyage/Day12-Chat-GPT/solvePart1.py
class BoxSection:
    def __init__(self, id, contains, originally_unknown):
        self.id = id
        self.contains = contains
        self.originally_unknown = originally_unknown

class Item:
    def __init__(self, id, length, placement_status):
        self.id = id
        self.length = length
        self.placement_status = placement_status

class BoxAndItemsRecord:
    def __init__(self, box, items):
        self.box = box
        self.items = items

def parse_input(input_string):
    rows = input_string.strip().split('\n')
    box_and_items_records = []

    for row in rows:
        box_data, items_list = row.split()

        # Parse box
        box_sections = []
        for i, char in enumerate(box_data):
            if char == '.':
                box_sections.append(BoxSection(i, "buffer material", False))
            elif char == '#':
                box_sections.append(BoxSection(i, "empty", False))
            elif char == '?':
                box_sections.append(BoxSection(i, "unknown", True))

        # Parse items
        items_data = items_list.split(',')
        items = [Item(i, int(length), "unplaced") for i, length in enumerate(items_data)]

        box_and_items_record = BoxAndItemsRecord(box_sections, items)
        box_and_items_records.append(box_and_items_record)

    return box_and_items_records

def read_input_from_file(filename):
    with open(filename, 'r') as file:
        return file.read()

def place_item_permanently(item, box, sequence):
    item.placement_status = "permanently placed"
    for box_section in sequence:
        box_section.contains = item.id

    for adjacent_section in get_adjacent_sections(sequence, box):
        if adjacent_section.contains == "unknown":
            adjacent_section.contains = "buffer material"

def place_item_temporarily(item, box, sequence):
    item.placement_status = "temporarily placed"
    for box_section in sequence:
        box_section.contains = item.id

    for adjacent_section in get_adjacent_sections(sequence, box):
        if adjacent_section.contains == "unknown":
            adjacent_section.contains = "buffer material"

def unplace_temporarily_placed_item(item, box, sequence):
    item.placement_status = "unplaced"
    for box_section in sequence:
        if box_section.originally_unknown:
            box_section.contains = "unknown"
        else:
            box_section.contains = "empty"

    for adjacent_section in get_adjacent_sections(sequence, box):
        if adjacent_section.originally_unknown:
            adjacent_section.contains = "unknown"

def get_adjacent_sections(sequence, box):
    adjacent_sections = []
    for box_section in sequence:
        left_adjacent = box_section.id - 1
        right_adjacent = box_section.id + 1

        if left_adjacent >= 0:
            adjacent_sections.append(box[left_adjacent])

        if right_adjacent < len(box):
            adjacent_sections.append(box[right_adjacent])

    return adjacent_sections

def find_potential_sequences(length, box):
    potential_sequences = []
    current_sequence = []

    for box_section in box:
        for box_section in box:
            if box_section.contains == "empty" or box_section.contains == "unknown":
                current_sequence.append(box_section)

                if len(current_sequence) == length:
                    potential_sequences.append(current_sequence)
                    break  # Add only the first potential sequence
            else:
                current_sequence = []

    return potential_sequences

def place_items_in_unambiguous_placements(record):
    box = record.box
    items = record.items

    while True:
        unplaced_items = [item for item in items if item.placement_status == "unplaced"]
        if not unplaced_items:
            break

        selected_item = max(unplaced_items, key=lambda x: x.length)

        potential_sequences = find_potential_sequences(selected_item.length, box)

        if len(potential_sequences) == 1:
            place_item_permanently(selected_item, box, potential_sequences[0])
        elif len(potential_sequences) > 1:
            same_length_items = [item for item in items if item.length == selected_item.length and item.placement_status == "unplaced"]

            if len(potential_sequences) == len(same_length_items):
                for item, sequence in zip(same_length_items, potential_sequences):
                    place_item_permanently(item, box, sequence)

    return record

def find_number_of_possible_arrangements(record):
    box = record.box
    items = record.items

    number_of_possible_arrangements = 0

    if all(item.placement_status == "permanently placed" for item in items):
        return 1

    for item in items:
        if item.placement_status == "unplaced":
            temporarily_place_item(item, box)
            number_of_possible_arrangements += find_number_of_possible_arrangements(record)
            unplace_temporarily_placed_item(item, box)

    return number_of_possible_arrangements

def main():
    total_number_of_possible_arrangements = 0

    file_input = read_input_from_file("testInput.dat")
    parsed_records = parse_input(file_input)

    for record in parsed_records:
        place_items_in_unambiguous_placements(record)
        total_number_of_possible_arrangements += find_number_of_possible_arrangements(record)

    print(f"Part 1: The number of possible arrangements is {total_number_of_possible_arrangements}")
    return total_number_of_possible_arrangements

if __name__ == "__main__":
    main()

File: 2023-Xtreme-Xmas-Code-Maiden-Voyage/Day12-Chat-GPT/solvePart1.ts
class BoxSection {
  id: number;
  contains: string;
  originally_unknown: boolean;

  constructor(id: number, contains: string, originally_unknown: boolean) {
    this.id = id;
    this.contains = contains;
    this.originally_unknown = originally_unknown;
  }
}

class Item {
  id: number;
  length: number;
  placement_status: string;

  constructor(id: number, length: number, placement_status: string) {
    this.id = id;
    this.length = length;
    this.placement_status = placement_status;
  }
}

class BoxAndItemsRecord {
  box: BoxSection[];
  items: Item[];

  constructor(box: BoxSection[], items: Item[]) {
    this.box = box;
    this.items = items;
  }
}

function parseInput(inputString: string): BoxAndItemsRecord[] {
  const rows = inputString.trim().split("\n");
  const boxAndItemsRecords: BoxAndItemsRecord[] = [];

  for (const row of rows) {
    const [boxData, itemsList] = row.split(" ");

    // Parse box
    const boxSections: BoxSection[] = [];
    for (let i = 0; i < boxData.length; i++) {
      const char = boxData[i];
      if (char === ".") {
        boxSections.push(new BoxSection(i, "buffer material", false));
      } else if (char === "#") {
        boxSections.push(new BoxSection(i, "empty", false));
      } else if (char === "?") {
        boxSections.push(new BoxSection(i, "unknown", true));
      }
    }

    // Parse items
    const itemsData = itemsList.split(",");
    const items: Item[] = itemsData.map((length, i) =>
      new Item(i, parseInt(length), "unplaced")
    );

    const boxAndItemsRecord = new BoxAndItemsRecord(boxSections, items);
    boxAndItemsRecords.push(boxAndItemsRecord);
  }

  return boxAndItemsRecords;
}

function readInputFromFile(filename: string): string {
  return Deno.readTextFileSync(filename);
}

function placeItemPermanently(
  item: Item,
  box: BoxSection[],
  sequence: BoxSection[],
): void {
  item.placement_status = "permanently placed";
  for (const boxSection of sequence) {
    boxSection.contains = item.id.toString();
  }

  for (const adjacentSection of getAdjacentSections(sequence, box)) {
    if (adjacentSection.contains === "unknown") {
      adjacentSection.contains = "buffer material";
    }
  }
}

function placeItemTemporarily(
  item: Item,
  box: BoxSection[],
  sequence: BoxSection[],
): void {
  item.placement_status = "temporarily placed";
  for (const boxSection of sequence) {
    boxSection.contains = item.id.toString();
  }

  for (const adjacentSection of getAdjacentSections(sequence, box)) {
    if (adjacentSection.contains === "unknown") {
      adjacentSection.contains = "buffer material";
    }
  }
}

function unplaceTemporarilyPlacedItem(
  item: Item,
  box: BoxSection[],
  sequence: BoxSection[],
): void {
  item.placement_status = "unplaced";
  for (const boxSection of sequence) {
    if (boxSection.originally_unknown) {
      boxSection.contains = "unknown";
    } else {
      boxSection.contains = "empty";
    }
  }

  for (const adjacentSection of getAdjacentSections(sequence, box)) {
    if (adjacentSection.originally_unknown) {
      adjacentSection.contains = "unknown";
    }
  }
}

function getAdjacentSections(
  sequence: BoxSection[],
  box: BoxSection[],
): BoxSection[] {
  const leftAdjacent = sequence[0].id - 1;
  const rightAdjacent = sequence[sequence.length - 1].id + 1;

  const adjacentSections: BoxSection[] = [];

  if (leftAdjacent >= 0) {
    adjacentSections.push(box[leftAdjacent]);
  }

  if (rightAdjacent < box.length) {
    adjacentSections.push(box[rightAdjacent]);
  }

  return adjacentSections;
}

function findPotentialSequences(
  length: number,
  box: BoxSection[],
): BoxSection[][] {
  const potentialSequences: BoxSection[][] = [];
  let currentSequence: BoxSection[] = [];

  for (const boxSection of box) {
    for (const boxSection of box) {
      if (
        boxSection.contains === "empty" || boxSection.contains === "unknown"
      ) {
        currentSequence.push(boxSection);

        if (currentSequence.length === length) {
          potentialSequences.push(currentSequence);
          break; // Add only the first potential sequence
        }
      } else {
        currentSequence = [];
      }
    }
  }

  return potentialSequences;
}

function placeItemsInUnambiguousPlacements(
  record: BoxAndItemsRecord,
): BoxAndItemsRecord {
  const box = record.box;
  const items = record.items;
  let itemPlacedThisSequence = true;

  while (itemPlacedThisSequence) {
    itemPlacedThisSequence = false;

    const unplacedItems = items.filter((item) =>
      item.placement_status === "unplaced"
    );

    if (unplacedItems.length === 0) {
      break;
    }

    const selectedItem = unplacedItems.reduce((maxItem, currentItem) =>
      currentItem.length > maxItem.length ? currentItem : maxItem
    );

    const potentialSequences = findPotentialSequences(selectedItem.length, box);

    if (potentialSequences.length === 1) {
      placeItemPermanently(selectedItem, box, potentialSequences[0]);
      itemPlacedThisSequence = true;
    } else if (potentialSequences.length > 1) {
      const sameLengthItems = items.filter(
        (item) =>
          item.length === selectedItem.length &&
          item.placement_status === "unplaced",
      );

      if (potentialSequences.length === sameLengthItems.length) {
        for (
          const [item, sequence] of sameLengthItems.map((
            item,
            index,
          ) => [item, potentialSequences[index]])
        ) {
          placeItemPermanently(item, box, sequence);
          itemPlacedThisSequence = true;
        }
      }
    }
  }

  return record;
}

function findNumberOfPossibleArrangements(record: BoxAndItemsRecord): number {
  const box = record.box;
  const items = record.items;

  let numberOfPossibleArrangements = 0;

  if (items.every((item) => item.placement_status === "permanently placed")) {
    return 1;
  }

  for (const item of items) {
    if (item.placement_status === "unplaced") {
      placeItemTemporarily(item, box);
      numberOfPossibleArrangements += findNumberOfPossibleArrangements(record);
      unplaceTemporarilyPlacedItem(item, box);
    }
  }

  return numberOfPossibleArrangements;
}

function main(): number {
  let totalNumberOfPossibleArrangements = 0;

  const fileInput = readInputFromFile("testInput.dat");
  const parsedRecords = parseInput(fileInput);

  for (const record of parsedRecords) {
    placeItemsInUnambiguousPlacements(record);
    totalNumberOfPossibleArrangements += findNumberOfPossibleArrangements(
      record,
    );
  }

  console.log(
    `Part 1: The number of possible arrangements is ${totalNumberOfPossibleArrangements}`,
  );
  return totalNumberOfPossibleArrangements;
}

if (import.meta.main) {
  main();
}
